cdap-app-fabric/src/main/java/co/cask/cdap/config/PreferencesStore.java

```diff
@@ -34,10 +42,12 @@
   private static final String EMPTY_NAMESPACE = "";
 
   private final ConfigStore configStore;
+  private final ProfileStore profileStore;
 
   @Inject
-  public PreferencesStore(ConfigStore configStore) {
+  public PreferencesStore(ConfigStore configStore, ProfileStore profileStore) {
```

Comment:
Strictly speaking using two store would result in incorrect logic, since each store would have their own transaction. This is similar to CDAP-13296. 

However, this case is slightly more tricker since the operations involves reading from one dateset (the profile dataset) to check for valid state, you gotta make that reading causes conflict with other transaction that is deleting the same profile.

Reply:
Actually why we need this PR? If someone assigned a disabled profile to a program, what kind of problem it can cause? Asking this because it would be quite difficult to handle the race condition during the app deployment / schedule creation.

Reply:
Since there can be race condition when deleting the profile. We want to have two restrictions when we delete a profile, 1. profile is disabled, 2. profile is not associated with any entities. If we allow entity to associate with a disabled profile, we will need to update the association, then we may delete the profile assuming there is no association(since the association update is not complete yet)

Reply:
I thought we said the reverse index is updated together with the preference in the same transaction, hence such race shouldn’t exists (or more precisely would cause tx conflict), isn’t it?

Reply:
If we allowing assigning while disabled, there is no race if we check the preference store and delete the profile in the same transaction. So although it would be better to do everything in the same tx, practically it doesn't actually matter if there is a race here and we assign a disabled profile to a program, since the check to delete the profile will catch that. 

One of the purposes of disabling is to allow an admin to 'lock' a profile, then systematically go through and remove it from pipelines, wait for current runs to finish, then delete the profile. So I still think we shouldn't allow assigning a disabled profile to a program at a conceptual level, but we don't need to be as careful with the race conditions during app deployment and schedule creation.

Reply:
That being said, I don't think the Store is the right abstraction layer to put the logic in, as it kind of implies everything is transactional and consistent. I think it would be better to have a PreferencesService that handles these types of things, especially since you'll need to add TMS publishing later too.